File: OrangeJuice.Server.Core/Services/AwsArgumentBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;

using OrangeJuice.Server.Configuration;
using OrangeJuice.Server.Data.Models;

namespace OrangeJuice.Server.Services
{
	public sealed class AwsArgumentBuilder : IArgumentBuilder
	{
		#region Fields

		private readonly AwsOptions _awsOptions;
		private readonly IDateTimeProvider _dateTimeProvider;
		#endregion

		#region Ctor
		public AwsArgumentBuilder(AwsOptions awsOptions, IDateTimeProvider dateTimeProvider)
		{
			_awsOptions = awsOptions;
			_dateTimeProvider = dateTimeProvider;
		}
		#endregion

		#region IArgumentBuilder members
		public IDictionary<string, string> BuildArgs(ProductDescriptorSearchCriteria searchCriteria)
		{
			DateTime now = _dateTimeProvider.GetNow();
			string timestamp = _dateTimeProvider.Format(now);

			// Ordering parameters in naturual byte order as required by AWS
			return new SortedDictionary<string, string>(StringComparer.OrdinalIgnoreCase)
			{
				{ "Operation", searchCriteria.Operation },
				{ "SearchIndex", searchCriteria.SearchIndex },
				{ "ResponseGroup", String.Join(",", searchCriteria.ResponseGroups ?? Enumerable.Empty<string>()) },
				{ "IdType", searchCriteria.IdType },
				{ "ItemId", searchCriteria.ItemId },
				{ "AWSAccessKeyId", _awsOptions.AccessKey },
				{ "AssociateTag", _awsOptions.AssociateTag },
				{ "Service", "AWSECommerceService" },
				{ "Condition", "All" },
				{ "Timestamp", timestamp }
			};
		}
		#endregion
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/AwsArgumentBuilder.fs
namespace OrangeJuice.Server.FSharp.Services

open System
open System.Collections.Generic
open System.Linq

open OrangeJuice.Server
open OrangeJuice.Server.Configuration
open OrangeJuice.Server.Data.Models
open OrangeJuice.Server.Services

type AwsArgumentBuilder(awsOptions : AwsOptions, dateTimeProvider : IDateTimeProvider) =
    interface IArgumentBuilder with
        member this.BuildArgs(searchCriteria : AwsProductSearchCriteria) : IDictionary<string, string> =
            let now = dateTimeProvider.GetNow()
            let timestamp = dateTimeProvider.Format(now)
            let group =
                if searchCriteria.ResponseGroups <> null
                    then searchCriteria.ResponseGroups :> IEnumerable<string>
                    else Enumerable.Empty<string>()
            [
                ("Operation", searchCriteria.Operation); 
                ("SearchIndex", searchCriteria.SearchIndex);
                ("ResponseGroup", String.Join(",", group));
                ("IdType", searchCriteria.IdType);
                ("ItemId", searchCriteria.ItemId);
                ("AWSAccessKeyId", awsOptions.AccessKey);
                ("AssociateTag", awsOptions.AssociateTag);
                ("Service", "AWSECommerceService");
                ("Condition", "All");
                ("Timestamp", timestamp)
            ] |> Map.ofList :> IDictionary<string, string>
